ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_dev/ExplicitDisclosureIT.scala

```diff
@@ -0,0 +1,421 @@
+        None,
+        checkDefiniteAnswerMetadata = true,
+      )
+    }
+  })
+
+  test(
+    "EDExerciseByKeyDisclosedContract",
+    "A disclosed contract can be exercised by key by non-stakeholder if authorized",
+    partyAllocation = allocate(TwoParties),
+    enabled = _.explicitDisclosure,
+  ) { implicit ec =>
+    { case Participants(Participant(ledger, owner, divulgee)) =>
+      for {
+        // Create contract with `owner` as only stakeholder
+        _ <- ledger.create(owner, WithKey(owner))
+        withKeyTxIds <- ledger.flatTransactionsByTemplateId(WithKey.id, owner)
+        withKeyCreate = createdEvents(withKeyTxIds(1)).head
+        withKeyDisclosedContract = createEventToDisclosedContract(withKeyCreate)
+        // Assert that a random party can exercise the contract by key (if authorized)
+        // when passing the disclosed contract to the submission
+        _ <- ledger.submitAndWait(
+          exerciseWithKey_byKey_request(ledger, owner, divulgee, withKeyDisclosedContract)
+        )
+      } yield ()
+    }
+  }
+
+  test(
+    "EDMetadata",
+    "All create events have metadata defined",
+    allocate(Parties(2)),
+    enabled = _.explicitDisclosure,
+  )(implicit ec => { case Participants(Participant(ledger, owner, delegate)) =>
+    val contractKey = ledger.nextKeyId()
+    for {
+      _ <- ledger.create(owner, Delegated(owner, contractKey))
+      _ <- ledger.create(owner, Delegation(owner, delegate))
+      flats <- ledger.flatTransactions(owner)
+      trees <- ledger.transactionTrees(owner)
+      someTransactionId = flats.head.transactionId
+      flatById <- ledger.flatTransactionById(someTransactionId, owner)
+      treeById <- ledger.transactionTreeById(someTransactionId, owner)
+    } yield {
+      assertLength("flatTransactions", 2, flats)
+      assertLength("transactionTrees", 2, trees)
+      assert(
+        flats.map(createdEvents).forall(_.forall(_.metadata.isDefined)),
+        "Metadata is empty for flatTransactions",
+      )
+      assert(
+        trees.map(createdEvents).forall(_.forall(_.metadata.isDefined)),
+        "Metadata is empty for transactionTrees",
+      )
+      assert(
+        createdEvents(flatById).forall(_.metadata.isDefined),
+        "Metadata is empty for flatTransactionById",
+      )
+      assert(
+        createdEvents(treeById).forall(_.metadata.isDefined),
+        "Metadata is empty for transactionTreeById",
+      )
+    }
+  })
+
+  test(
+    "EDArchivedDisclosedContracts",
+    "The ledger rejects archived disclosed contracts",
+    allocate(Parties(2)),
+    enabled = _.explicitDisclosure,
+  )(implicit ec => { case Participants(Participant(ledger, owner, delegate)) =>
+    for {
+      testContext <- initializeTest(ledger, owner, delegate)
+
+      // Archive the disclosed contract
+      _ <- ledger.exercise(owner, testContext.delegatedCid.exerciseArchive())
+
+      // Exercise the choice using the now inactive disclosed contract
```

Comment:
Consider adding a test that this does succeed if the contract is not actually used.

Reply:
Good point!